timemachine/cpp/src/bd_exchange_move.cu

```diff
@@ -100,28 +131,89 @@ void BDExchangeMove<RealType>::move_device(
         curandErrchk(templateCurandUniform(cr_rng_, d_translations_.data, d_translations_.length));
 
         sampler_.sample_device(num_target_mols_, num_samples, d_log_weights_before_.data, d_samples_.data, stream);
+
+        k_setup_sample_atoms<<<ceil_divide(num_samples, tpb), tpb, 0, stream>>>(
+            num_samples,
+            mol_size_,
+            d_samples_.data,
+            d_atom_idxs_.data,
+            d_mol_offsets_.data,
+            d_target_mol_atoms_.data,
+            d_target_mol_offsets_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        k_atom_by_atom_energies<<<atom_by_atom_grid, tpb, 0, stream>>>(
+            N,
+            mol_size_,
+            d_target_mol_atoms_.data,
+            d_coords,
+            d_params_.data,
+            d_box,
+            nb_beta_,
+            cutoff_squared_,
+            d_sample_per_atom_energy_buffer_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
+        // Subtract off the weights for the individual waters from the sampled water
+        // clobbers the sampled mol energy value, but we have to special case that anyways
+        k_adjust_weights<RealType, true><<<ceil_divide(num_target_mols_, tpb), tpb, 0, stream>>>(
+            N,
+            num_target_mols_,
+            mol_size_,
+            d_atom_idxs_.data,
+            d_mol_offsets_.data,
+            d_sample_per_atom_energy_buffer_.data,
+            beta_, // 1 / kT
+            d_log_weights_after_.data);
+        gpuErrchk(cudaPeekAtLastError());
+
         k_rotate_and_translate_mols<RealType><<<ceil_divide(num_samples, tpb), tpb, 0, stream>>>(
             num_samples,
             d_coords,
             d_box,
             d_samples_.data,
-            d_mol_offsets_.data,
+            d_target_mol_offsets_.data,
             d_quaternions_.data,
             d_translations_.data,
             d_intermediate_coords_.data);
         gpuErrchk(cudaPeekAtLastError());
 
-        mol_potential_.mol_energies_device(
+        k_atom_by_atom_energies<<<atom_by_atom_grid, tpb, 0, stream>>>(
```

Comment:
Not urgent, but I wonder if we can combine the two `k_atom_by_atom_energies` later on (to lessen kernel overhead etc.), i.e. given `old_mol_pos`, `new_mol_pos`, and `other_coords`, (we'd only have to load in `other_coords` once as opposed to twice)

Reply:
We could combine these two calls into a single call